packages/babel-core/src/config/files/module-types.ts

```diff
@@ -151,92 +164,104 @@ export default function* loadCodeDefault(
           // fall through: require() failed due to TLA
         } else if (
           e.code === "ERR_REQUIRE_ESM" ||
-          (!process.env.BABEL_8_BREAKING && ext === ".mjs")
+          (!process.env.BABEL_8_BREAKING && type === "esm")
         ) {
           // fall through: require() failed due to ESM
         } else {
           throw e;
         }
       }
     // fall through: require() failed due to ESM or TLA, try import()
-    case "auto .mjs":
+    case "auto esm":
       if ((async ??= yield* isAsync())) {
-        return (yield* waitFor(loadMjsFromPath(filepath))).default;
+        const promise = isTS
+          ? ensureTsSupport(filepath, ext, () => loadMjsFromPath(filepath))
+          : loadMjsFromPath(filepath);
+
+        return (yield* waitFor(promise)).default;
       }
       throw new ConfigError(esmError, filepath);
     default:
       throw new Error("Internal Babel error: unreachable code.");
   }
 }
 
-function loadCtsDefault(filepath: string) {
-  const ext = ".cts";
-  const hasTsSupport = !!(
+function ensureTsSupport<T>(
+  filepath: string,
+  ext: string,
+  callback: () => T,
+): T {
+  if (
     require.extensions[".ts"] ||
     require.extensions[".cts"] ||
     require.extensions[".mts"]
-  );
+  ) {
+    return callback();
+  }
 
-  let handler: NodeJS.RequireExtensions[""];
+  if (ext !== ".cts") {
+    throw new ConfigError(
+      `\
+You are using a ${ext} config file, but Babel only supports transpiling .cts configs. Either:
```

Comment:
This error is "new", but before we would try to parse the files as JSON and thus just give a worse error.